pymc/step_methods/state.py

```diff
@@ -0,0 +1,98 @@
+#   Copyright 2024 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+from copy import deepcopy
+from dataclasses import dataclass, fields
+from typing import Any
+
+import numpy as np
+
+
+class MetaDataClassState(type):
+    def __new__(cls, name, bases, namespace):
+        return dataclass(eq=False, kw_only=True)(super().__new__(cls, name, bases, namespace))
+
+
+class DataClassState(metaclass=MetaDataClassState):
+    def __eq__(self, other: Any) -> bool:
+        if self.__class__ != other.__class__:
+            return False
+        this_fields = set([f.name for f in fields(self)])
+        other_fields = set([f.name for f in fields(other)])
+        for field in this_fields:
+            self_val = getattr(self, field)
+            other_val = getattr(other, field)
+            if not DataClassState.compare_values(self_val, other_val):
+                return False
+        return this_fields == other_fields
+
+    @staticmethod
+    def compare_values(v1: Any, v2: Any) -> bool:
+        if v1.__class__ != v2.__class__:
+            return False
+        if isinstance(v1, (list, tuple)):  # noqa: UP038
+            return len(v1) == len(v2) and all(
+                DataClassState.compare_values(v1i, v2i) for v1i, v2i in zip(v1, v2)
+            )
+        elif isinstance(v1, dict):
+            if set(v1) != set(v2):
+                return False
+            return all(DataClassState.compare_values(v1[k], v2[k]) for k in v1)
+        elif isinstance(v1, np.ndarray):
+            return bool(np.array_equal(v1, v2, equal_nan=True))
+        elif isinstance(v1, np.random.Generator):
+            return DataClassState.compare_values(v1.bit_generator.state, v2.bit_generator.state)
+        else:
+            return v1 == v2
+
+
+class WithSamplingState:
```

Comment:
Can you add docstrings for the three new classes to explain how they fit together?

For `WithSamplingState` I understand that it's a mixin adding a `sampling_state` property which, upon access, returns a new container object of a `DataClassState` subtype. This container holds copies of field values of the `WithSamplingState` object. (?)

Reply:
That's exactly right. I'll add comments and docstrings.